sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ForLoopConditionAlwaysFalse.cs

```diff
@@ -0,0 +1,218 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class ForLoopConditionAlwaysFalse : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S2252";
+        private const string MessageFormat = "This loop will never execute.";
+
+        private static readonly ISet<SyntaxKind> ConditionsToCheck = new HashSet<SyntaxKind>
+        {
+            SyntaxKind.GreaterThanExpression,
+            SyntaxKind.GreaterThanOrEqualExpression,
+            SyntaxKind.LessThanExpression,
+            SyntaxKind.LessThanOrEqualExpression,
+            SyntaxKind.EqualsExpression,
+            SyntaxKind.NotEqualsExpression
+        };
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var forNode = (ForStatementSyntax)c.Node;
+                    if (forNode.Condition != null && (IsAlwaysFalseCondition(forNode.Condition) || IsConditionFalseAtInitialization(forNode)))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, forNode.Condition.GetLocation()));
+                    }
+                },
+                SyntaxKind.ForStatement);
+        }
+
+        private bool IsAlwaysFalseCondition(ExpressionSyntax condition) =>
+            condition.IsKind(SyntaxKind.FalseLiteralExpression) ||
+                (
+                    IsLogicalNot(condition, out var logicalNode) &&
+                    IsAlwaysTrueCondition(logicalNode.Operand.RemoveParentheses()
+                 ));
+
+        private bool IsAlwaysTrueCondition(ExpressionSyntax condition) =>
+            condition.IsKind(SyntaxKind.TrueLiteralExpression) ||
+                (
+                    IsLogicalNot(condition, out var logicalNode) &&
+                    IsAlwaysFalseCondition(logicalNode.Operand.RemoveParentheses())
+                );
+
+        private static bool IsLogicalNot(ExpressionSyntax expression, out PrefixUnaryExpressionSyntax logicalNot)
```

Comment:
instance `private` methods are mixed with `private static` methods - could you please put the static ones first, and then the non-static ones?